ports-of-call/portable_errors.hpp

```diff
@@ -0,0 +1,198 @@
+// on its behalf a nonexclusive, paid-up, irrevocable worldwide license
+// in this material to reproduce, prepare derivative works, distribute
+// copies to the public, perform publicly and display publicly, and to
+// permit others to do so.
+// ========================================================================================
+
+// C++ headers
+#include <iostream>
+#include <stdexcept>
+
+// C headers
+#include <cstdio>
+#include <cstdlib>
+
+// TODO(JMM): Is relative path right here?
+#include "portability.hpp"
+
+#define PORTABLE_ALWAYS_REQUIRE(condition, message)                            \
+  if (!(condition)) {                                                          \
+    PortsOfCall::ErrorChecking::require(#condition, message, __FILE__,         \
+                                        __LINE__);                             \
+  }
+
+#define PORTABLE_ALWAYS_REQUIRE_THROWS(condition, message)                     \
+  if (!(condition)) {                                                          \
+    PortsOfCall::ErrorChecking::require_throws(#condition, message, __FILE__,  \
+                                               __LINE__);                      \
+  }
+
+#define PORTABLE_ALWAYS_ABORT(message)                                         \
+  PortsOfCall::ErrorChecking::abort(message, __FILE__, __LINE__);
+
+#define PORTABLE_ALWAYS_THROW(message)                                         \
+  PortsOfCall::ErrorChecking::abort_throws(message, __FILE__, __LINE__);
+
+#define PORTABLE_ALWAYS_WARN(message)                                          \
+  PortsOfCall::ErrorChecking::warn(message, __FILE__, __LINE__);
+
+#ifdef PORTABILITY_STRATEGY_NONE
+#define PORTABLE_ALWAYS_THROW_IFEXCEPT(message) PORTABLE_ALWAYS_THROW(message)
+#else
+#define PORTABLE_ALWAYS_THROW_IFEXCEPT(message) PORTABLE_ALWYAS_ABORT(message)
+#endif // PORTABILITY_STRATEGY_NONE
+
+#fdef NDEBUG
+#define PORTABLE_REQUIRE(condition, message) ((void)0)
+#define PORTABLE_REQUIRE_THROWS(condition, message) ((void)0)
+#define PORTABLE_ABORT(message) ((void)0)
+#define PORTABLE_THROW(message) ((void)0)
+#define PORTABLE_WARN(message) ((void)0)
+#define PORTABLE_THROW_IFEXCEPT(message) ((void)0)
+#else
+#define PORTABLE_REQUIRE(condition, message)                                   \
+  PORTABLE_ALWAYS_REQUIRE(condition, message)
+#define PORTABLE_REQUIRE_THROWS(condition, message)                            \
+  PORTABLE_ALWAYS_REQUIRE_THROWS(condition, message)
+#define PORTABLE_ABORT(message) PORTABLE_ALWAYS_ABORT(message)
+#define PORTABLE_THROW(message) PORTABLE_ALWAYS_THROW(message)
+#define PORTABLE_WARN(message) PORTABLE_ALWAYS_WARN(message)
+#define PORTABLE_THROW_IFEXCEPT(message) PORTABLE_ALWAYS_THROW_IFEXCEPT(message)
+#endif // NDEBUG
+
+namespace PortsOfCall {
+namespace ErrorChecking {
+namespace impl {
+[[noreturn]] KOKKOS_INLINE_FUNCTION void abort(const char *const message) {
+#ifdef PORTABILITY_STRATEGY_KOKKOS
```

Comment:
Shouldn't the `KOKKOS_INLINE_FUNCTION` be behind the `ifdef`?

Reply:
Or should it use `PORTABLE_FUNCTION`?

Reply:
It should just be `PORTABLE_INLINE_FUNCTION` my bad.